keps/0030-immutable-parameters.md

```diff
@@ -0,0 +1,76 @@
+---
+kep-number: 30
+short-desc: Immutable parameters that can only be set on installation
+title: Immutable parameters
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-04-28
+last-updated: 2020-04-28
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes a feature to allow parameters to be defined as immutable. Immutable parameters can not be updated after the installation of the operator
+
+## Motivation
+
+Certain aspects of an operator must be defined while installing an operator and can not be changed later on. 
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter can not be changed after the installation
+
+### Non-Goals
+
+- Parameters that can be set once after the installation - this KEP requires a parameter to be set in the installation process.
+
+## Proposal
+
+An flag `immutable` is added to the parameter definition. If that flag is true, the parameter needs to be set (or have a default) when the operator is installed. The parameter can not be updated with `kudo update -p` anymore.
+
+```yaml
+parameters:
+  - name: NUM_TOKENS
+    description: The number of tokens on each node of the cluster
+    default: 256
+    immutable: true
+```
+
+The default value for `immutable` is `false`.
+
+### User Stories
+
+- [#652](https://github.com/kudobuilder/kudo/issues/652) Make it possible to specify immutable parameter
+
+- NUM_TOKENS in the Cassandra operator
+- STORAGE_CLASS in the Zookeeper operator
+- DISK_SIZE, STORAGE_CLASS and PERSISTENT_STORAGE in the Kafka operator
+
+### Implementation Details/Notes/Constraints
+
+The KUDO CLI will check if a parameter is immutable when updating an instance and error out.
+
+The admission webhook should verify if the parameter is immutable and reject the update to the instance on update.
+
+
+Open Questions: 
+- Should it be allowed to use an immutable parameter in a `kudo update` invocation when the value is the same as in the instance? This would effectively not be an update. 
+  - Yes: The CLI would have to either check the current instance or rely on the admission webhook to error out
```

Comment:
```suggestion
  - Yes: The CLI would have to either check the current instance and operatorversion defaults or rely on the admission webhook to error out
```
we should check OV defaults values also, in case Instance didn't use any custom value

Reply:
Good point, there's an interesting edge-case here: If the Instance uses the default value from the OV, what happens when updating the OV and its default value changes? One way around this would be to always provide a parameter value in the Instance for immutable parameters, even when using the default from the OV.

Reply:
Oh yeah, that's a good point. I'll add that to the KEP.

I think both options have valid arguments:
1) Copy the default over as and actual value, so it is indeed immutable, even when the default changes in OV
2) Always using the default - when the OV default changes, assume the operator developer has thought about it

I think I prefer variant 1. Makes it harder to change things, but looking at the use cases, these things simply *must* not change after installation.

Reply:
You can't prevent users from editing the OV and running a `deploy` plan, right? Which is an argument in favor of (2) point. I'm ok with leaving this to the OV developer. We might provide some CLI safety to `kudo upgrade` e.g. `An immutable parameter has changed, are you really, really sure?`

Reply:
No, you can't prevent the users from editing the OV - that's why the idea 1 is to copy over the default from OV into I and not use the OV after the initial deploy. Of course users could disable the webhook and manually change the parameter in the instance - but at that point they hopefully know what they're doing ;)

I think that allows more safety for OV developers: If we were to always use the OV value, it might prevent operator developers from changing the defaults, otherwise it may break peoples installation when the upgrade.

Reply:
OV developer and Instance user don't have to be the same. Hence I'm favoring option 1. Changing defaults in OV can be fatal for running instances. We should err on the side of caution here.

Reply:
I agree, side of caution seems the best way from my side as well - I've updated the KEP itself already to take that into account.